### DISCUSSION TITLE
Config framework

### DISCUSSION THREAD

**Discussion by chris-langfield** [ComputationalCryoEM/ASPIRE-Python#698] [Ideas]
Some things we want to be able to do with a config system after talking with @garrettwrong:
1 - Automatically detect config file in package directory. In particular, we want automatic path searching. For instance, first look for a `config.ini` in the working directory, then in the home directory, finally defaulting to a config file shipped with the package.
2 - A simple file format. We currently use `.ini` along with the standard library's `configparser`. Other options are YAML, TOML, or even `.py` 
3 - We may want an environment variable set to path to config file?
4-  We want to be able to ship a GPU config file ready to go

Some notes on the options out there:
- [`configparser`](https://docs.python.org/3/library/configparser.html): From what I can tell, this does not implement 1) above. It could be quite involved to write a cross-platform path search
- [`confuse`](https://confuse.readthedocs.io/en/latest/): This is for YAML. We have heard good things from RC's Python expert
- [`dynaconf`](https://dynaconf.readthedocs.io/en/docs_223/): This is popular on Reddit and StackOverflow threads, but looks like it's probably overkill. Also seems to have partly dropped 3.7 support
- [`config-path`](https://github.com/barry-scott/config-path) This might do what we want as far as cross-platform path search. It hasn't been updated in a while but looks official and is marked as "production/stable" on PyPI.

`configparser` + `config-path` + INI's is jumping out at me as the simplest solution to look into first.

**Comment by chris-langfield** [comment]
Rolling our own environment variable code seems easier to me than the cross-platform path search. Maybe we can still do this even if we use a package that doesn't implement env variable stuff. `configparser` doesn't seem to have it built in

**Comment by garrettwrong** [comment]
Hey Chris, thanks for putting this info together and asking our RSE friend.

I took a look at `confuse`.  I like the way they are marketing the overrides, but the API seemed a bit much for what we need... and I'd need to hunt for simple working examples, which is not a good sign.  I will look at `dynaconf`, since we might be off 3.7 soon enough anyway...

From our chat today, I think we confirmed we are just looking to config numerical backends and logging.  Everything else should probably be in our code/classes.  The logging potentially has a fair amount of config.

**Comment by janden** [comment]
Matplotlib approach: https://matplotlib.org/stable/tutorials/introductory/customizing.html

**Comment by garrettwrong** [comment]
Leaning towards `confuse`, though I would prefer it had more widespread adoption.  I will try to implement it this afternoon and we can see what it looks like.
